### ISSUE TITLE
Which mail_uid and mail_gid

### ISSUE THREAD

**Issue by Wermorl** [grosjo/fts-xapian#173]
Hi grosjo,

I was running fts-xapian for a long time (great tool thanks!) and just realised that any search on my inbox (but not other folders ) had not hits. After investigating the problem i found that fts-xapian is no longer able to do the indexing because of permission issues:

`doveadm index -A -q "foobar"`

gets me to this:

`doveadm(nobody)<5408><m5kPDtd2CWcgFQAAxfi2ig>: Error: mkdir(/nonexistent/Maildir/xapian-indexes) failed: Permission denied (euid=65534(nobody) egid=65534(nogroup))
doveadm(nobody)<5408><m5kPDtd2CWcgFQAAxfi2ig>: Error: FTS Xapian: can not create '/nonexistent/Maildir/xapian-indexes'
doveadm(nobody)<5408><m5kPDtd2CWcgFQAAxfi2ig>: Error: FTS Xapian: You need to set mail_uid and mail_gid in your dovecot.conf according to the user of mail_location (/nonexistent/Maildir)
doveadm(nobody)<5408><m5kPDtd2CWcgFQAAxfi2ig>: Error: fts: Failed to initialize backend 'xapian': 
doveadm(systemd-coredump)<5408><4G1dDtd2CWcgFQAAxfi2ig>: Error: mkdir(//Maildir/xapian-indexes) failed: Permission denied (euid=999(systemd-coredump) egid=999(systemd-coredump) missing +w perm: /, dir owned by 0:0 mode=0755)
doveadm(systemd-coredump)<5408><4G1dDtd2CWcgFQAAxfi2ig>: Error: FTS Xapian: can not create '//Maildir/xapian-indexes'
doveadm(systemd-coredump)<5408><4G1dDtd2CWcgFQAAxfi2ig>: Error: FTS Xapian: You need to set mail_uid and mail_gid in your dovecot.conf according to the user of mail_location (//Maildir)
doveadm(systemd-coredump)<5408><4G1dDtd2CWcgFQAAxfi2ig>: Error: fts: Failed to initialize backend 'xapian':`

So the errormessage is quite obvious, but how would I set the user in mail_uid and mail_gid as the Maildir folders are the users home dir and owned by those users.

What am I missing here?

Thanks for reading

**Comment by grosjo** [comment]
From your logs, I can see that your "mail_location" variable is not set properly

What is your Maildir location for each user ?

Refer to  https://doc.dovecot.org/2.3/configuration_manual/mail_location/

**Comment by grosjo** [comment]
Probably you shall set 
`mail_location = maildir:~/Maildir`
in your dovecot.conf

**Comment by grosjo** [comment]
Another info : the cron job is there just to cleanup the indexes, but not to index anything (which is done in real time)

So if you do not get indexed properly, maybe you shall update to latest relaease (1.7.16) and delete your "xapian-indexes" folder in each Maildir

**Comment by Wermorl** [comment]
Thanks for your help!

`mail_location = maildir:~/Maildir` is and was set like this.
Update to latest git was done
xapian-indexes folder is deleted

The error remains the same

**Comment by Wermorl** [comment]
Soe here is a strange one:

The errormessage for `doveadm index -A -q "foobar"` remains as above. Bit the index Folder is back again and indexing works like a charm .

What now?

**Comment by philmck** [comment]
Same problem here. In case it helps, I've noticed that the indexer-worker user sometimes matches the mailbox being indexed, but sometimes doesn't.

Here's an example of a complete mismatch (user "snugtoes" trying to reindex user "hayessymphony") resulting in failure:
`2024-10-15T19:27:24.886638+01:00 server dovecot: indexer-worker(bounces@snugtoes.com)<3372263><oAWsMAy0DmfndDMAXOOVSg>: Error: FTS Xapian: Can't re-create Xapian DB /home/hayessymphony/homes/treasurer/Maildir/xapian-indexes/db_c7032d05b74a5f6521041c00652dd778 : DatabaseCreateError - Permission denied
`

**Comment by grosjo** [comment]
@philmck do you have userdb setup properly ?
https://doc.dovecot.org/2.3/configuration_manual/authentication/user_databases_userdb/

mine looks like this
userdb {
        args = /etc/mail/dovecot/dovecot-sql.conf
        driver = sql
}

**Comment by grosjo** [comment]
Feel free to re-open if you need help

**Comment by Wermorl** [comment]
I Updated to latest git and tried again. The error remains on both systems i use. One is debian 12.7 and one Ubuntu 20.04.

Search still works and indexing too as it seems.

**Comment by grosjo** [comment]
But canyou share your dovecot.conf file ?

**Comment by Wermorl** [comment]
So I found a thing. 

The errormessages are all for users which do not have a home directory.

For example:

`nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin
`

**Comment by Wermorl** [comment]
I found this because i accidentally removed a homefolder which belomged to a still existing user. And afterwards i got the same errormessage but for the missing home folder.

**Comment by grosjo** [comment]
please share your config of dovecot (doveconf -n)

**Comment by Wermorl** [comment]
Thanks for looking into this. Here is the config:

```
# 2.3.7.2 (3c910f64b): /etc/dovecot/dovecot.conf
# Pigeonhole version 0.5.7.2 ()
# OS: Linux 5.4.0-193-generic x86_64 Ubuntu 20.04.6 LTS 
# Hostname: cloud
default_vsz_limit = 4 G
mail_fsync = never
mail_location = maildir:~/Maildir
mail_plugins = " fts fts_xapian"
mail_privileged_group = mail
managesieve_notify_capability = mailto
managesieve_sieve_capability = fileinto reject envelope encoded-character vacation subaddress comparator-i;ascii-numeric relational regex imap4flags copy include variables body enotify environment mailbox date index ihave duplicate mime foreverypart extracttext
namespace inbox {
  inbox = yes
  location = 
  mailbox Drafts {
    special_use = \Drafts
  }
  mailbox Junk {
    special_use = \Junk
  }
  mailbox Sent {
    special_use = \Sent
  }
  mailbox "Sent Messages" {
    special_use = \Sent
  }
  mailbox Trash {
    special_use = \Trash
  }
  prefix = 
}
passdb {
  driver = pam
}
plugin {
  fts = xapian
  fts_autoindex = yes
  fts_autoindex_exclude = \Trash
  fts_enforced = yes
  fts_xapian = partial=3 full=20 attachments=0 verbose=0
  sieve = file:~/sieve;active=~/.dovecot.sieve
}
postmaster_address = postmaster@example.com
protocols = " imap sieve sieve"
service imap-login {
  inet_listener imap {
    port = 143
  }
  inet_listener imaps {
    port = 993
    ssl = yes
  }
}
ssl_cert = </opt/dehydrated/certs/path-to-file/fullchain.pem
ssl_dh = # hidden, use -P to show it
ssl_key = # hidden, use -P to show it
userdb {
  driver = passwd
}
protocol lmtp {
  mail_fsync = optimized
}
protocol lda {
  mail_fsync = optimized
  mail_plugins = " fts fts_xapian sieve"
}
protocol imap {
  mail_max_userip_connections = 20
}
local_name mail.domainone.com {
  ssl_cert = </opt/dehydrated/certs/path-to-file/fullchain.pem
  ssl_key = # hidden, use -P to show it
}
local_name mail.domaintwo.com {
  ssl_cert = </opt/dehydrated/certs/path-to-other-file/fullchain.pem
  ssl_key = # hidden, use -P to show it
}
```

**Comment by grosjo** [comment]
Maybe chec this https://doc.dovecot.org/2.3/configuration_manual/authentication/user_databases_userdb/

You may need to add a "userdb" setting to set the user properly

**Comment by Wermorl** [comment]
Thanks! I will look into this.

**Comment by philmck** [comment]
It looks like Webmin has NOT set up a userdb. In that case, why does
    Xapian indexing ever work? And if it works most of the time, why
    does it sometimes fail?
    
    Also, if I do set up this userdb, how will it be populated and
    maintained? Is it a manual task or will it happen automagically? 
    
    I had a look at the documentation but I'm none the wiser.
    
    Thanks for the suggestion, in any case.
    
    On 2024-12-08 13:35, Joan Moreau wrote:
    
    
      
      Maybe chec this https://doc.dovecot.org/2.3/configuration_manual/authentication/user_databases_userdb/
      You may need to add a "userdb" setting to set the
        user properly
      —
        Reply to this email directly, view it on GitHub, or unsubscribe.
        You are receiving this because you were mentioned.Message
          ID: ***@***.***>
      [
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/grosjo/fts-xapian/issues/173#issuecomment-2525910459",
"url": "https://github.com/grosjo/fts-xapian/issues/173#issuecomment-2525910459",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by grosjo** [comment]
Where do you configure your users initially ?

**Comment by philmck** [comment]
Sorry, I've had to disable fts-xapian for now because it seems to be causing mail delivery to freeze, which is creating delivery problems (and repeated alerts from uptimerobot). Also searching mail bodies from an IMAP client was working but still taking too long to be practical (minutes). It's very intermittent though and it's not at all obvious to me how dovecot could be interfering with postfix, so there may be more than one issue here. Because of the intermittent nature it's going to take days to isolate the culprit with any certainty.

Looking at the comments in the dovecot mailing list it seems this is a known problem that could be fixed by using the latest version (which would require compiling it myself instead of using the Ubuntu package). Is that correct?

I don't know the answer to your question "where do you configure your users?" I'm afraid. The short answer is in Virtualmin but that may not be helpful because there's no documentation about where they are stored.

**Comment by philmck** [comment]
Would disabling the workers be a workaround?

**Comment by grosjo** [comment]
I mean, the users of dovecot , you must configure them somewhere in your dovecot.conf (userdb / passwd paragraph ) , right ? What driver/parameter do you use ?

Which version of the plugin are you using ?

**Comment by philmck** [comment]
Really struggling to find a userdb, I'm afraid.

Dovecot -n shows:
```
userdb {
  driver = passwd
}
and 
passdb {
  driver = pam
}
```
if that's any help.

I would have thought 10-auth.conf would give me the answer but it's almost empty and almost everything in auth-system.conf.ext is commented out except this:

```
userdb {
  # <doc/wiki/AuthDatabase.Passwd.txt>
  driver = passwd
  # [blocking=no]
  #args = 

  # Override fields from passwd
  #override_fields = home=/home/virtual/%u
}
```
I haven't been able to find a user database in the postfix configuration either. (There are some simple text files holding aliases and virtual addresses.)

The xapian package name is dovecot-fts-xapian 1.6.0-1 which probably tells us something about the version.

It's getting very late here, I might be able to make some more sense of it in the morning.

**Comment by philmck** [comment]
There's also this package installed:
libxapian30 1.4.22-1build1

**Comment by philmck** [comment]
I found the virtualmin create_user() code used for adding dovecot users in the attached file starting at line #1439 but it's huge.
[virtual-server-lib-funcs.zip](https://github.com/user-attachments/files/18309121/virtual-server-lib-funcs.zip)

**Comment by grosjo** [comment]
No, I was referring to dovecot. How do you administer dovecot users ? Have you set the reference of users db (see the docs https://doc.dovecot.org/2.3/configuration_manual/authentication/user_databases_userdb/ ) to connect to your "virtualmin"  software ?

**Comment by philmck** [comment]
That's it. That's how I edit dovecot users - the GUI (or the Virtualmin API) is all I have. Virtualmin > [select site] > Edit Users. I know of no other way. I don't know where the control panel stores them "under the hood" because it's not documented. I'll need to ask them or work it out from the source code.

**Comment by philmck** [comment]
Update: It looks like they're all system users in /etc/shadow. Where exactly this is configured in dovecot I don't know (all I can find is commented out lines so I'm guessing it just uses defaults). Obviously it works because users are getting their emails.

**Comment by philmck** [comment]
All the userdb settings in Dovecot seem to be commented out, so is it just using that by default?

**Comment by philmck** [comment]
Update: There are two uncommented lines in /etc/dovecot/conf.d/auth-system.conf.ext, spotted by someone in the Virtualmin support forum. They agree with my doveconf -n extracts above.

**Comment by grosjo** [comment]
Are you using latest git ?

**Comment by grosjo** [comment]
> I found this because i accidentally removed a homefolder which belomged to a still existing user. And afterwards i got the same errormessage but for the missing home folder.

@Wermorl  If your users have no home folder, where do you store their emails ?

**Comment by philmck** [comment]
> Are you using latest git ?

Are you asking me? If so, we're going in circles because I posted the versions of the apt packages I'm using above, and above that asked whether I need to use the latest git version in order to get the fix I need. If there's a better way of providing information, please spell out what I need to do to find get the answers you need.

Maybe a new thread is needed, because I think there may be several different bugs involved here.

**Comment by grosjo** [comment]
Yes, you must use the latest git , so we compare apples with apples

**Comment by philmck** [comment]
OK, I was afraid of that (because it typically takes days).

So far I have failed at the first two steps. :-(

At https://github.com/grosjo/fts-xapian it says 

> You will need to configure properly [Users Home Directories](https://wiki.dovecot.org/VirtualUsers/Home) in dovecot configuration

but that link says 
> The old wiki has been closed

Never mind, I know that users have home directories that work.

I'm on Ubuntu, so the next step is:
`apt-get build-dep dovecot-core`

but that just returns

> E: You must put some 'deb-src' URIs in your sources.list

How do I do that?

**Comment by grosjo** [comment]
https://askubuntu.com/questions/1512042/ubuntu-24-04-getting-error-you-must-put-some-deb-src-uris-in-your-sources-list

or just type
`sed -i 's/^Types: deb$/Types: deb deb-src/' /etc/apt/sources.list.d/ubuntu.sources`

**Comment by philmck** [comment]
Thanks, it seems to have compiled OK now (at least in a VM, I'm not game to try it on a production server yet). How do I check the fts-xapian version?

**Comment by philmck** [comment]
Also: Can I just copy the compiled plugin to the production server? (I realise apt might overwrite it in future.)

**Comment by grosjo** [comment]
> Thanks, it seems to have compiled OK now (at least in a VM, I'm not game to try it on a production server yet). How do I check the fts-xapian version?
> 

just make sure "make install" returns no errors and the new .so file is in dovecot modules folders (/usr/lib/dovecot/modules or similar)

restart dovecot

**Comment by grosjo** [comment]
> Also: Can I just copy the compiled plugin to the production server? (I realise apt might overwrite it in future.)
> 

you rather need to compile it on the prod server, as some libraries may differ

**Comment by grosjo** [comment]
any progress ?

**Comment by philmck** [comment]
I now have (potentially) two different versions of fts-xapian running, how do I find the version number to confirm which one is active?

**Comment by grosjo** [comment]
You can not have 2 versions running on the same dovecot, dovecot does not allow it

**Comment by grosjo** [comment]
I added the version number in latest git (see at loading time in the log)

**Comment by philmck** [comment]
OK, we're inching towards a solution. I'm searching the log without success so far. According to https://doc.dovecot.org/2.3/admin_manual/logging/ (and "doveadm log find") dovecot logs to syslog and mail.log but searching those for "fts-xapian" shows nothing. Yet I know the entries exist somewhere because logwatch finds them. Confused and out of time today, I'll have another go later.

**Comment by grosjo** [comment]
Should search for "xapian" only in the logs

**Comment by philmck** [comment]
Update: I've now compiled the latest git code on a production server (running Ubuntu 24.04.1). It wasn't easy. Still not seeing a xapian version number anywhere (but found some past successes and failures in the mail log). It will probably take a few days to know if it has made a difference.

**Comment by seb1k** [comment]
Yeap, it could be nice to have this kind of functionality :


```
root@server:~# doveadm fts info
fts plugin 1.8.x
xapian version 1.x.y
indexed mailbox 45
indexes size 56G

... why not something else ...
```

Anyway, good plugin :)

**Comment by philmck** [comment]
Update: The latest version from git seems to have broken the search entirely. :-(

I'm seeing errors like this in mail.log:

`2025-01-14T15:23:44.469144+00:00 london dovecot: imap(phil@mckerracher.net)<1301334><ZkNrJawrXs4qBfSAEAAJmlQABf/+LQg+>: FTS Xapian: 'BSO' (/home/mckerracher/homes/phil/Maildir/xapian-indexes/db_4c75b6341c954563130200000f3e080a) indexes do not exist. Initializing DB
2025-01-14T15:23:44.056887+00:00 london dovecot: imap(phil@mckerracher.net)<1301334><ZkNrJawrXs4qBfSAEAAJmlQABf/+LQg+>: Warning: FTS Xapian: 'BSO' (/home/mckerracher/homes/phil/Maildir/xapian-indexes/db_4c75b6341c954563130200000f3e080a_exp.db) expunge does not exist. Creating it
2025-01-14T15:23:42.613749+00:00 london dovecot: imap(phil@mckerracher.net)<1301334><ZkNrJawrXs4qBfSAEAAJmlQABf/+LQg+>: Warning: FTS Xapian: 'BSO' (/home/mckerracher/homes/phil/Maildir/xapian-indexes/db_4c75b6341c954563130200000f3e080a_dict.db) dictionnary does not exist. Creating it`

If I try even a simple search for words that I know exist in the subject line of emails in my inbox I now get no results.

**Comment by philmck** [comment]
That's weird. When it says "Creating it", the location of the xapian-indexes folder looks correct but it now seems to be actually deleting it instead of creating it.

**Comment by grosjo** [comment]
Are you really using latest git ? (try "git pull" in the folder you are using to compile, then "make install"

**Comment by philmck** [comment]
I only did the pull yesterday and it compiled something so probably yes.

"git pull" shows errors:

```
# git pull
remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 15 (delta 8), reused 15 (delta 8), pack-reused 0 (from 0)
Unpacking objects: 100% (15/15), 14.42 KiB | 1.20 MiB/s, done.
From https://github.com/grosjo/fts-xapian
 + 875c426...540112e master     -> origin/master  (forced update)
 * [new tag]         1.8.5      -> 1.8.5
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint:
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint:
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.
```

**Comment by philmck** [comment]
```
# make install
Making install in src
make[1]: Entering directory '/root/fts-xapian/src'
make[2]: Entering directory '/root/fts-xapian/src'
make[2]: Nothing to be done for 'install-exec-am'.
 /usr/bin/mkdir -p '/usr/lib/dovecot/modules'
 /bin/bash ../libtool   --mode=install /usr/bin/install -c   lib21_fts_xapian_plugin.la '/usr/lib/dovecot/modules'
libtool: install: /usr/bin/install -c .libs/lib21_fts_xapian_plugin.so /usr/lib/dovecot/modules/lib21_fts_xapian_plugin.so
libtool: install: /usr/bin/install -c .libs/lib21_fts_xapian_plugin.lai /usr/lib/dovecot/modules/lib21_fts_xapian_plugin.la
libtool: install: /usr/bin/install -c .libs/lib21_fts_xapian_plugin.a /usr/lib/dovecot/modules/lib21_fts_xapian_plugin.a
libtool: install: chmod 644 /usr/lib/dovecot/modules/lib21_fts_xapian_plugin.a
libtool: install: ranlib /usr/lib/dovecot/modules/lib21_fts_xapian_plugin.a
libtool: finish: PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/sbin" ldconfig -n /usr/lib/dovecot/modules
----------------------------------------------------------------------
Libraries have been installed in:
   /usr/lib/dovecot/modules

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the '-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the 'LD_RUN_PATH' environment variable
     during linking
   - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to '/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
make[2]: Leaving directory '/root/fts-xapian/src'
make[1]: Leaving directory '/root/fts-xapian/src'
make[1]: Entering directory '/root/fts-xapian'
make[2]: Entering directory '/root/fts-xapian'
make[2]: Nothing to be done for 'install-exec-am'.
make[2]: Nothing to be done for 'install-data-am'.
make[2]: Leaving directory '/root/fts-xapian'
make[1]: Leaving directory '/root/fts-xapian'
root@london.beeches.it:~/fts-xapian#

```

**Comment by philmck** [comment]
In case it's relevant I do see occasional memory warnings like this in the mail log:

`2025-01-14T08:11:53.563031+00:00 london xapian-docswriter: DW #3 (Sent,/home/mckerracher/homes/amanda/Maildir/xapian-indexes/db_b8a3d924b4520f5f8e2f00000f3e080a) - Committing 5001 docs due to low free memory (1772 MB vs 300 MB) or Cached docs exceeded (5001 vs 5000 limit)`

By "occasional" I mean only a minority of mailboxes are affected, but the error seems consistent for the affected ones.

**Comment by philmck** [comment]
Also just in case it's relevant, I found one instance of the original "You need to set mail_uid and mail_gid" error in the mail.log from yesterday.

`2025-01-13T08:16:15.201225+00:00 london dovecot: imap(phil@mckerracher.net)<16229><wgTiDpIrAIvUhPQO>: Error: FTS Xapian: You need to set mail_uid and mail_gid in your dovecot.conf according to the user of mail_location (/home/mckerracher/homes/phil/Maildir)`

This may be a red herring though because it's a rare occurrence (the only one in the log) and the server was rebooted yesterday due to an unrelated problem at the datacentre.

**Comment by grosjo** [comment]
hi

To cleanup your setup, can you execute the follwoing commands (as root)

```
cd /root
rm -rf fts-xapian
git clone https://github.com/grosjo/fts-xapian.git
cd fts-xapian
autoreconf -vi --force
./configure --with-dovecot=/usr/lib/dovecot
make clean
make
make install
```

Once this is done , restart dovecot
```
systemctl restart dovecot
```
and then follow-up the logs
```
journalctl -f | grep -i xapian
```

**Comment by philmck** [comment]
Done, I'll let you know the log results.

**Comment by philmck** [comment]
Output attached, in case it helps...note the warning on line 194

> fts-backend-xapian-functions.cpp:1529:56: warning: too many arguments for format [-Wformat-extra-args]
>  1529 |                                         syslog(LOG_ERR,"FTS Xapian: Can not search keyword : %s",sql.c_str(),zErrMsg);
>       |                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

[results.txt](https://github.com/user-attachments/files/18418070/results.txt)

**Comment by grosjo** [comment]
please type " git pull" again

**Comment by philmck** [comment]
Done.

> root@london.beeches.it:~/fts-xapian# git pull
> remote: Enumerating objects: 17, done.
> remote: Counting objects: 100% (17/17), done.
> remote: Compressing objects: 100% (6/6), done.
> remote: Total 17 (delta 11), reused 17 (delta 11), pack-reused 0 (from 0)
> Unpacking objects: 100% (17/17), 1.88 KiB | 137.00 KiB/s, done.
> From https://github.com/grosjo/fts-xapian
>    21b1889..11257fd  master     -> origin/master
>  * [new tag]         1.8.6      -> 1.8.6
> Updating 21b1889..11257fd
> Fast-forward
>  Makefile.am                          |  4 ++--
>  configure.ac                         |  2 +-
>  src/Makefile.am                      |  7 +------
>  src/fts-backend-xapian-functions.cpp | 31 ++++++++++---------------------
>  src/fts-backend-xapian.cpp           | 10 ++++------
>  src/fts-xapian-plugin.h              |  8 ++++----
>  6 files changed, 22 insertions(+), 40 deletions(-)

(Also "make install" and "systemctl restart dovecot". Nothing "xapian" in logs so far.)

**Comment by grosjo** [comment]
You need to know where your logs are sent to. 
Recent systems uses journald , so you can see logs via "journalctl -f"

**Comment by philmck** [comment]
Sorry, I wasn't clear - there are "xapian" entries in the logs, there just hadn't been any new ones since the change above.

Some have now arrived, however. They mainly look like this:

> 2025-01-16T03:03:39.901227+00:00 london dovecot: imap(ian@mckerracher.net)<198104><zHNfCsorzg4BkawF>: FTS Xapian: 'INBOX' (/home/mckerracher/homes/ian/Maildir/xapian-indexes/db_52bf7c278fa8485b487900000f3e080a) indexes do not exist. Initializing DB
> 2025-01-16T03:03:39.897973+00:00 london dovecot: imap(ian@mckerracher.net)<198104><zHNfCsorzg4BkawF>: Warning: FTS Xapian: 'INBOX' (/home/mckerracher/homes/ian/Maildir/xapian-indexes/db_52bf7c278fa8485b487900000f3e080a_exp.db) expunge does not exist. Creating it
> 2025-01-16T03:03:39.877325+00:00 london dovecot: imap(ian@mckerracher.net)<198104><zHNfCsorzg4BkawF>: Warning: FTS Xapian: 'INBOX' (/home/mckerracher/homes/ian/Maildir/xapian-indexes/db_52bf7c278fa8485b487900000f3e080a_dict.db) dictionnary does not exist. Creating it
> 2025-01-16T03:03:39.874192+00:00 london dovecot: imap(ian@mckerracher.net)<198104><zHNfCsorzg4BkawF>: Warning: FTS Xapian: 'INBOX' new Version of the pluging (1.8.6)

It seems that everything is being reindexed. That may be good news, I guess.

**Comment by grosjo** [comment]
these are just saying that plugin have been upgraded and indeed, mailboxes are being re-indexed, as this lead to a large saving of space

**Comment by philmck** [comment]
It seems to have finished reindexing after I ran **doveadm index -A -q \*** but I'm not seeing the results I expect. I used Roundcube to search my local mail folder (using IMAP, but on the same server). As a test, I searched for the word "China". I expected about 186 hits because that's what Thunderbird finds in the same mail folder on my PC.

![China search](https://github.com/user-attachments/assets/4aa3763f-d884-4bfb-8d81-a856dbd5452b)

The xapian search using Roundcube found only 13 hits, all in 2022. It also took 25 seconds, which is an improvement over the several minutes it took before, but still a bit disappointing. The missing results are my main concern, though. Any idea what's going wrong?

**Comment by grosjo** [comment]
Maybe Thunderbird has downloaded the email on your local PC, hasn't he ? (and therefore deleted form the distant dovecot server)

**Comment by grosjo** [comment]
or are you using the same filters (in roundcube, you need to specify which fields are to be searched)

**Comment by grosjo** [comment]
![Image](https://github.com/user-attachments/assets/2f408db9-efec-4a5e-96a5-22e65cfedd0b)

**Comment by philmck** [comment]
Thanks for the help but Thunderbird is not deleting messages from the server and I have tried selecting "Body" and "Entire Message" in Roundcube.

The symptoms I'm seeing seem to indicate that IMAP search (using xapian) still isn't working. I'll try some more clients to see if I can narrow it down but there's certainly something still missing.

On the plus side, I haven't seen any new xapian errors in the log. And I love the new reduced database size! The old databases were costing me money because they were nearly as big as the mailboxes they were indexing.

**Comment by philmck** [comment]
Update: Full text search seems to work from Edison Mail on a phone, but only one folder at a time. That's good news!

I tried to solve the "one folder at a time" issue by creating a virtual "All Mail" folder earlier and it sort of worked, but it caused problems with eM Client so I had to disable it again.

So, there are two outstanding questions, both related to the IMAP search protocol. They probably need separate threads.

1. Why doesn't FTS work from Roundcube? Has anyone found a configuration that does work?
2. Why doesn't "All Mail" FTS work with eM Client? Again, has anyone found a configuration that does work?

A more general question is, has anyone found **any** IMAP FTS client compatible with fts-xapian that works from phones, searching across folders, including the bodies of messages and going back more than a year? 

How do you test fts-xapian? You must have some favourite mail apps that you use for testing. If we knew how it was tested we would at least have more confidence that it works some of the time.

**Comment by grosjo** [comment]
1 - FTS works very well from Roundcube, I am myself using roundcube on a daily basis and all is fine. However, Roundcube is much less " foggy" than Thunderbird as it just searches exactly where you ask him to search

2 - There may be a bug in the "multi folder search" -> please open a new " issue" here on github

If no more issues with "ownership" of the indexes, please lets' close this issue

**Comment by grosjo** [comment]
@Wermorl can you confirm from your side if latest version still raises concerns on folder ownership ?

**Comment by philmck** [comment]
Thanks for the Roundcube info, I'll experiment further.

I already reported the "multi folder" issue elsewhere but it's a Dovecot or eM Client problem, not specific to Xapian I think.

I suspect the folder permission fix is only temporary because it will presumably be overwritten the next time the apt fts-xapian package updates. Do I need to uninstall that? Can I uninstall it without breaking everything?

**Comment by grosjo** [comment]
The next time the apt gets updated, it will be with the latest version, most probably

No need to "uninstall" anything

So, Isuggest to close this issue , except is @Wermorl is still experiencing the issue

**Comment by philmck** [comment]
I'm still seeing lots of errors in the logs but they're now "indexes do not exist" errors rather than "permission denied". I hope they will all disappear after a couple of days. It was always intermittent anyway.

Update: Those log errors are now 24 hours old, no mention at all of "xapian" today.

**Comment by philmck** [comment]
Still no luck with Roundcube. It's returning a very small number of results and taking a long time to do so - I'm not sure it's connecting to Xapian at all. What's the secret?

Update: It's not just Roundcube. I have still not found **any** mail app for an Android phone that will do a full text search of my emails (including the message bodies, in any folder, going back more than a year). Nor have I found a Webmail client that can do this from a browser. So I'm stuck.

**Comment by grosjo** [comment]
if the search is taking a very long time, it means dovecot is not using any plugin for search.

can you please dump "doveconf -n" here ?

**Comment by philmck** [comment]
Sure...

[doveconf_result.txt](https://github.com/user-attachments/files/18462473/doveconf_result.txt)

**Comment by grosjo** [comment]
and do you see any error at loading (hall ve something like "Starting FTS 
Xapian v1.8.5 xxxx"

**Comment by philmck** [comment]
Hmm, no. The last "xapian" messages in the logs are these and similar from two days ago. Running **systemctl restart dovecot** shows no problems (and mail is being delivered) but no log messages about xapian now at all.

> 
2025-01-16T04:27:09.116660+00:00 london dovecot: indexer-worker(philip@crestcape.com)<597921><GA++Bp2KiGehHwkAcMVIOA>: Warning: FTS Xapian: 'Heritage House.BT' new Version of the pluging (1.8.6)

**Comment by philmck** [comment]
I'm getting the same result on two other servers. I can give you SSH access if that would help. My phone is going to be off for most of the next 24 hours so responses may be a bit slow, sorry.

**Comment by grosjo** [comment]
Ys, curious to access via SSH to see what creates the pb

**Comment by grosjo** [comment]
My ssh key : AAAAB3NzaC1yc2EAAAADAQABAAABAQC7fM/a/DONp/IduuotIm/bsie1V7Mn6J23Ecr2zYID+JJJdm4AjwI9IXfFOMD4kD55g/bBCF4x1FdMcouBlR/11PSWQx+4r/eBQZg7i8R2hB2rGG9M44zNJLyZEz5fDtaBfz3g
SiKBHD0dtGqAps/nuILknxgvDbCvWu43Me8VuR0tXLidG9EQcI/fOvGoykTxS9JEBYxMyIN7kslBClZnyCgZhJI24UR4EuphR9zxRXuPKbAu0Fxh/+q8tqkBqHXz5I97OZ8Bpdsl2HOFNVM/VHU8gX1I/5iVXqFaicEPSgJCugbz
DWS+HS5dGFhFpmMtkNIfU06wiVB/chykclAz

**Comment by philmck** [comment]
Key added, server is london.beeches.it.

**Comment by grosjo** [comment]
it seems xapian is properly installed (cf /var/log/mail.log)
(I set verbose=1 to enable log lines)

2025-01-19T00:17:08.250320+00:00 london dovecot: imap(ian@mckerracher.net)<1398143><UEVqEAQspo8kBi1ARxuSEJyCVy2cmGGo>: FTS Xapian: Starting version 1.8.6 with partial=3 verbose=1 max_threads=3 lowmemory=300 MB

**Comment by grosjo** [comment]
Found this

```
doveadm(_imunify): Error: mkdir(/home/_imunify/Maildir/xapian-indexes) failed: Permission denied (euid=995(_imunify) egid=985(_imunify) missing +w perm: /home, dir owned by 0:0 mode=0755)
doveadm(_imunify): Error: FTS Xapian: can not create '/home/_imunify/Maildir/xapian-indexes'
doveadm(_imunify): Error: FTS Xapian: You need to set mail_uid and mail_gid in your dovecot.conf according to the user of mail_location (/home/_imunify/Maildir)

```

but the pb is that /home/_imunify does not even exists ! You need to create the proper home directories of your users before receiving/indexing mail, if you use Maildir

**Comment by philmck** [comment]
That's correct - there are several users on the server that don't have home directories or mailboxes, this is normal. Dovecot (and previous versions of xapian, I think) ignores them.

In case it's relevant though, that particular user has been added quite recently (because ImunifyAV has only recently implemented support for Ubuntu 24.04).

**Comment by grosjo** [comment]
everything works fine. Can you launch a search from any mailbox that " takes forever"  ?

**Comment by philmck** [comment]
I've just searched phil@mckerracher.net for the word "China" (including body text, all folders) and it took 45 seconds to return results that only seem to include messages with "China" in the subject. I used webmail.beeches.it (Roundcube).

**Comment by grosjo** [comment]
it seems indeed that xapian is not called

**Comment by grosjo** [comment]
test

**Comment by grosjo** [comment]
I can not see any email being indexed on that server. It seems tehre is a "procmail"  process that redirect all the email to another server

**Comment by grosjo** [comment]
indeed , in postfix.conf, you have

`mailbox_command = /usr/bin/procmail-wrapper -o -a $DOMAIN -d $LOGNAME
`

basically, emails are not handled by dovecot for reception, only for reading

**Comment by grosjo** [comment]
Suggest you clean-up your setup to remove all the errors and inconsistencies, including webmin, so you have a proper setup

**Comment by grosjo** [comment]
lmtp is not even configured on dovecot

**Comment by grosjo** [comment]
see https://doc.dovecot.org/2.3/configuration_manual/howto/postfix_dovecot_lmtp/

**Comment by philmck** [comment]
OK, I think it's time to give up. Mail and Webmin are working fine apart from FTS, replacing them is not an option (too expensive). Thanks for all your efforts, it's appreciated.

**Comment by philmck** [comment]
Is there an easy way to remove all the indexes to reclaim the disk space?

**Comment by grosjo** [comment]
You can delete the xapian-indexes folders. 

But you need anyway to configure your server properly

**Comment by philmck** [comment]
I've spoken to the Virtualmin people and there's nothing "improper" about filtering mail with procmail before passing it to dovecot. So I don't think there's going to be any change there, in case you thought it was just a mistake.
